comp110-worksheet-6/OxoBoard.cs

```diff
@@ -7,55 +7,270 @@
+			//Will return Mark for the specific board cell
+			return board[x, y];
 		}
 
 		// If the specified square is currently empty, fill it with mark and return true.
 		// If the square is not empty, leave it as-is and return False.
 		public bool SetSquare(int x, int y, Mark mark)
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+			//Try and catch are used to stop values that are more than or less than the size of the grid.
+			try
+			{
+				if (board[x, y] == Mark.None)
+				{
+					//Assigns the board cell to the given mark paramater and returns true.
+					board[x, y] = mark;
+					return true;
+				}
+				else
+				{
+					//Return false if the square is already occupied.
+					return false;
+				}
+			}
+			catch
+			{
+				//Returns false if an error would've occurred, this happens when the index is out of range
+				return false;
+			}
 		}
 
 		// If there are still empty squares on the board, return false.
 		// If there are no empty squares, return true.
 		public bool IsBoardFull()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+			//Nested for loop for searching for empty board cells, uses GetLongLength() to find length of x and y axis
+			for (int i = 0; i < board.GetLongLength(0); i++)
+			{
+				for (int j = 0; j < board.GetLongLength(1); j++)
+				{
+					if(board[i,j] == Mark.None)
+					{
+						return false;
+					}
+				}
+			}
+			return true;
 		}
 
 		// If a player has three in a row, return Mark.O or Mark.X depending on which player.
 		// Otherwise, return Mark.None.
 		public Mark GetWinner()
 		{
-			throw new NotImplementedException("TODO: implement this function and then remove this exception");
+			//Here nested for loops are used to search through all combinations of the cells to determine if a player
+			//Has won or lost the game, this checks through the y axis.
+			for (int horizontal = 0; horizontal < board.GetLongLength(0); horizontal++)
+			{
+				int inCommon = 1;
+				for (int vertical = 1; vertical < board.GetLongLength(1); vertical++)
+				{
+					//Checks to see if the next board position is equal to the current board position
+					if (board[horizontal, vertical] == board[horizontal, vertical - 1])
+					{
+						inCommon++;
+						//In common is used to keep track of the current line of same marks on the board.
+						if(inCommon >= numForARow)
+						{
+							//If the cells in common are player marks, it will return the player's mark
+							if(board[horizontal, vertical] != Mark.None)
+							{
+								return board[horizontal, vertical];
+							}
+						}
+					}
+					else
+					{
+						//Resets the search back to the beginning if a board position does not follow the current pattern.
+						inCommon = 1;
+					}
+				}
+			}
+			//Checks for any winners along the x axis
+			for(int vertical = 0; vertical < board.GetLongLength(1); vertical++)
```

Comment:
This code is quite similar to the horizontal check -- could they have been refactored into a single helper function?